tun/tun_solaris.go

```diff
@@ -0,0 +1,398 @@
+
+		if uintptr(n) < unsafe.Sizeof(unix.IfMsghdr{}) {
+			continue
+		}
+		msg := (*unix.IfMsghdr)(unsafe.Pointer(&data[0]))
+		if int(msg.Type) != unix.RTM_IFINFO {
+			continue
+		}
+		if int(msg.Index) != tunIfindex {
+			continue
+		}
+
+		iface, err := net.InterfaceByName(tun.name)
+		if err != nil {
+			tun.errors <- err
+			return
+		}
+
+		// Currently the solaris/illumos tun driver has two quirks that should be noted here
+		// 1. At creation time it immediately goes up before this code will have
+		// a chance to notice
+		// 2. It doesn't really respect 'ifconfig tunN down' properly so this code
+		// won't notice, and it will be broken and refuse to come back up
+		// if it is set down.
+		// If that ever changes, this code will be waiting.
+
+		// Up / Down event
+		up := (iface.Flags & net.FlagUp) != 0
+		if up != statusUp && up {
+			tun.events <- EventUp
+		}
+		if up != statusUp && !up {
+			tun.events <- EventDown
+		}
+		statusUp = up
+
+		// MTU changes
+		if iface.MTU != statusMTU {
+			tun.events <- EventMTUUpdate
+		}
+		statusMTU = iface.MTU
+	}
+}
+
+func tunDestroy(name string) error {
+	var l unix.Lifreq
+	if err := l.SetName(name); err != nil {
+		return err
+	}
+	// cleanup both ipv4 and ipv6
+	for _, proto := range []int{unix.AF_INET, unix.AF_INET6} {
+		fd, err := unix.Socket(proto, unix.SOCK_DGRAM, 0)
+		if err != nil {
+			return fmt.Errorf("could not open Socket of type %v", proto)
+		}
+		defer unix.Close(fd)
+
+		// get the STREAMS muxid for the tun device stream
+		reqnum := int(unix.SIOCGLIFMUXID)
+
+		// solaris/illumos defines the ioctl number as a signed int, but the
+		// common x/sys/unix functions use uint.
+		if err := unix.IoctlLifreq(fd, uint(reqnum), &l); err != nil {
+			continue
+		}
+
+		id := l.GetLifruInt()
+
+		// Unlink the tun stream from the IP mutiplexor
+		if err := unix.IoctlSetInt(fd, unix.I_PUNLINK, id); err != nil {
+			continue
+		}
+	}
+	return nil
+}
+
+func CreateTUN(name string, mtu int) (Device, error) {
+	ppa := -1
+	if name != "tun" {
+		_, err := fmt.Sscanf(name, "tun%d", &ppa)
+		if err != nil || ppa < 0 {
+			return nil, fmt.Errorf("interface name must be tun[0-9]*")
+		}
+	}
+
+	fd, err := unix.Open(tunNode, unix.O_RDWR, 0)
+	if err != nil {
+		return nil, fmt.Errorf("could not open TUN (%s)", tunNode)
+	}
+
+	// The "tun" device is a STREAMS module.  We need to make a STREAMS ioctl
+	// request to that module using the TUNNEWPPA command.  Returns the newly
+	// allocated PPA number. Passing -1 will request a dynamically assinged
+	// PPA. Passing a specific id will try to use the one specified.
+	var s unix.Strioctl
+	s.Cmd = unix.TUNNEWPPA
+	s.SetInt(ppa)
+	ppa, err = unix.IoctlSetStrioctlRetInt(fd, unix.I_STR, &s)
+	if err != nil {
+		return nil, err
+	}
+
+	assignedName := fmt.Sprintf("tun%d", ppa)
+
+	// Clean up anything possibly left behind by a crash
+	tunDestroy(assignedName)
```

Comment:
Interesting...

What have you experienced being left behind? What are the circumstances here?